File: 2/2_cuatrimestre/PDOO/practicas/practica 2/deepSpace_2/src/deepspace/SpaceStation.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package deepspace;

import java.util.ArrayList;


/**
 *
 * @author manbolq
 */
public class SpaceStation {
    private final int MAXFUEL = 100;
    private final float SHIELDLOSSPERUNITSHOT = 0.1f;
    
    private float ammoPower;
    private float fuelUnits;
    private String name;
    private int nMedals;
    private float shieldPower;
    private Hangar hangar;
    private ArrayList<ShieldBooster> shieldBoosters;
    private ArrayList<Weapon> weapons;
    private Damage pendingDamage;
    
    
    
    /**************************************************************************/
    /*************************    CONSTRUCTORES   *****************************/
    /**************************************************************************/
    
    SpaceStation(String n, SuppliesPackage supplies){
        name = n;
        ammoPower = supplies.getAmmoPower();
        fuelUnits = supplies.getFuelUnits();
        shieldPower = supplies.getShieldPower();
        weapons = new ArrayList<Weapon>();
        hangar = new Hangar(0);
        nMedals = 0;
        pendingDamage = new Damage(0, 0);
        shieldBoosters = new ArrayList<ShieldBooster>();
    }
    
    
    /**************************************************************************/
    /****************************    METODOS   ********************************/
    /**************************************************************************/
    
    private void assignFuelValue(float f){
        if (f <= MAXFUEL)
            fuelUnits = f;
    }
    
    private void cleanPendingDamage(){
        if (pendingDamage.hasNoEffect())
            pendingDamage = new Damage(0, 0);
    }
    
    public void cleanUpMountedItems(){
        for (int i = 0; i < weapons.size(); i++){
            if (weapons.get(i).getUses() == 0)
                weapons.remove(i);
        }
        
        for (int i = 0; i < shieldBoosters.size(); i++){
            if (shieldBoosters.get(i).getUses() == 0)
                shieldBoosters.remove(i);
        }
    }
    
    public void discardHangar(){
        hangar = new Hangar(0);
    }
    
    public void discardShieldBooster(int i){
        throw new UnsupportedOperationException();
    }
    
    public void discardShieldBoosterInHangar(int i){
        if (hangar != null)
            hangar.removeShieldBooster(i);
    }
    
    public void discardWeapon(int i){
        throw new UnsupportedOperationException();
    }
    
    public void discardWeaponInHangar(int i){
        if (hangar != null)
            hangar.removeWeapon(i);
    }
    
    public float fire(){
        throw new UnsupportedOperationException();
    }
    
    public float getAmmoPower(){
        return ammoPower;
    }
    
    public float getFuelUnits(){
        return fuelUnits;
    }
    
    public Hangar getHangar(){
        return hangar;
    }
    
    public String getName(){
        return name;
    }
    
    public int getNMedals(){
        return nMedals;
    }
    
    public Damage getPendingDamage(){
        return pendingDamage;
    }
    
    public ArrayList<ShieldBooster> getShieldBoosters(){
        return shieldBoosters;
    }
    
    public float getShieldPower(){
        return shieldPower;
    }
    
    public float getSpeed(){
        return fuelUnits/MAXFUEL;
    }
    
    public SpaceStationToUI getUIversion(){
        return new SpaceStationToUI(this);
    }
    
    public ArrayList<Weapon> getWeapons(){
        return weapons;
    }
    
    public void mountShieldBooster(int i){
        if (hangar != null){
            ShieldBooster aux = hangar.removeShieldBooster(i);
            if (aux != null)
                shieldBoosters.add(aux);
        }
    }
    
    public void mountWeapon(int i){
        if (hangar != null){
            Weapon aux = hangar.removeWeapon(i);
            if (aux != null)
                weapons.add(aux);
        }
    }
    
    public void move(){
        fuelUnits -= getSpeed();
        if (fuelUnits < 0)
            fuelUnits = 0;
    }
    
    
    public float protection(){
        throw new UnsupportedOperationException();
    }
    
    public void receiveHangar(Hangar h){
        if (hangar == null || hangar.getMaxElements() == 0)
            hangar = h;
    }
    
    public boolean receiveShieldBooster(ShieldBooster s){
        if (hangar != null)
            return hangar.addShieldBooster(s);
        else
            return false;
    }
    
    public ShotResult receiveShot(float shot){
        throw new UnsupportedOperationException();
    }
    
    public void receiveSupplies(SuppliesPackage s){
        ammoPower += s.getAmmoPower();
        shieldPower += s.getShieldPower();
        fuelUnits += s.getFuelUnits();
    }
    
    public boolean receiveWeapon(Weapon w){
        if (hangar != null)
            return hangar.addWeapon(w);
        else
            return false;
    }
    
    public void setLoot(Loot loot){
        throw new UnsupportedOperationException();
    }
    
    public void setPendingDamage(Damage d){
        pendingDamage = d.adjust(weapons, shieldBoosters);
    }
    
    public boolean validState(){
        return ((pendingDamage == null) || (pendingDamage.hasNoEffect()));
    }
    
    
    @Override
    public String toString(){
        return getUIversion().toString();
    }
    
    
    
}

File: 2/2_cuatrimestre/PDOO/practicas/practica 3/deepSpaceRuby/SpaceStation.rb
require_relative 'Damage'
require_relative 'Weapon'
require_relative 'ShieldBooster'
require_relative 'Hangar'
require_relative 'SpaceStationToUI'
require_relative 'ShotResult'
require_relative 'CardDealer'

module Deepspace
    class SpaceStation

        attr_reader :ammoPower, :fuelUnits, :weapons, :hangar, :shieldPower, :name, :nMedals, :pendingDamage, :shieldBoosters

        def initialize(n, supplies)
            @MAXFUEL = 100
            @SHIELDLOSSPERUNITSHOT = 0.1
            @name = n
            @ammoPower = supplies.ammoPower
            @fuelUnits = supplies.fuelUnits
            @shieldPower = supplies.shieldPower
            @weapons = Array.new
            @hangar = nil
            @nMedals = 0
            @pendingDamage = nil
            @shieldBoosters = Array.new
            
        end


        def cleanUpMountedItems
            for i in 0..(@weapons.length - 1)
                if (@weapons.at(i).uses == 0)
                    @weapons.delete_at(i)
                end
            end

            for i in 0..(@shieldBoosters.length-1)
                if (@shieldBoosters.at(i).uses == 0)
                    @shieldBoosters.delete_at(i)
                end
            end
        end

        def discardHangar
            @hangar = nil
        end

        def discardShieldBooster(i)
            size = @shieldBoosters.length

            if (i >= 0 && i < size)
                w = @shieldBoosters.remove(i)
                if(@pendingDamage != nil)
                    @pendingDamage.discardWeapon(w)
                    cleanPendingDamage
                end
            end
        end

        def discardShieldBoosterInHangar(i)
            if (@hangar != nil)
                @hangar.removeShieldBooster(i)
            end
        end

        def discardWeapon(i)
            size = @weapons.length

            if (i >= 0 && i < size)
                w = @weapons.delete_at(i)
                if(@pendingDamage != nil)
                    @pendingDamage.discardWeapon(w)
                    cleanPendingDamage
                end
            end
        end


        def discardWeaponInHangar(i)
            if (@hangar != nil)
                @hangar.removeWeapon(i)
            end
        end


        def fire
            size = @weapons.length
            factor = 1

            for i in 0..(size-1)
                factor *= @weapons[i].useIt
            end
            
            return @ammoPower*factor
        end

        def mountShieldBooster(i)
            if (@hangar != nil)
                aux = @hangar.removeShieldBooster(i)
                if (aux != nil)
                    shieldBoosters << aux
                end
            end
        end


        def mountWeapon(i)
            if (@hangar != nil)
                aux = @hangar.removeWeapon(i)
                if (aux != nil)
                    weapons << aux
                end
            end
        end


        def move
            @fuelUnits -= getSpeed
            if (@fuelUnits < 0)
                @fuelUnits = 0
            end
            
        end

        def protection
            size = @shieldBoosters.length
            factor = 1

            for i in 0..(size - 1)
                factor *= @shieldBoosters[i].useIt
            end

            return @shieldPower*factor
        end

        def receiveHangar(h)
            if (@hangar == nil)
                @hangar = h
            end

        end

        def receiveShieldBooster(s)

            if (@hangar != nil)
                @hangar.addShieldBooster(s)
            else
                false
            end
        end

        def receiveShot(shot)
            myProtection = protection

            if (myProtection >= shot)
                @shieldPower -= @SHIELDLOSSPERUNITSHOT*shot
                if (@shieldPower < 0)
                    @shieldPower = 0
                end
                return ShotResult::RESIST
            else
                @shieldPower = 0
                return ShotResult::DONOTRESIST
            end
        end


        def receiveSupplies(s)
            @ammoPower += s.ammoPower
            @shieldPower += s.shieldPower
            @fuelUnits += s.fuelUnits
        end


        def receiveWeapon(w)
            if (@hangar != nil)
                @hangar.addWeapon(w)
            else
                false
            end

        end

        def setLoot(loot)
            dealer = CardDealer.instance
            h = loot.nHangars
      
            if h > 0
              hangar = dealer.nextHangar
              receiveHangar(hangar)
            end
      
            elements = loot.nSupplies

            elements.times do
              sup = dealer.nextSuppliesPackage
              receiveSupplies(sup)
            end
      
            elements = loot.nWeapons

            elements.times do
              weap = dealer.nextWeapon
              receiveWeapon(weap)
            end
      
            elements = loot.nShields

            elements.times do
              sh = dealer.nextShieldBooster
              receiveShieldBooster(sh)
            end
      
            @nMedals += loot.nMedals
        end

        def getSpeed
            @fuelUnits/@MAXFUEL
        end

        def setPendingDamage(d)
            @pendingDamage = d.adjust(weapons, shieldBoosters)

        end


        def validState
            (@pendingDamage == nil) || (@pendingDamage.hasNoEffect)
        end


        def getUIversion
            return SpaceStationToUI.new(self)
        end

        def to_s 
            return getUIversion().to_s
        end

        
        private

        def cleanPendingDamage
            if (@pendingDamage.hasNoEffect)
                @pendingDamage = nil
            end


        end


        def assignFuelValue(f)
            if (f <= @MAXFUEL)
                @fuelUnits = f
            end

        end


    end
end

